Hazelcast.Net/Hazelcast.IO.Serialization/BufferPool.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+// http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Hazelcast.Core;
+
+namespace Hazelcast.IO.Serialization
+{
+    internal class BufferPoolThreadLocal : IDisposable
+    {
+        private readonly ISerializationService _serializationService;
+        private readonly ThreadLocal<WeakReference> _threadLocal = new ThreadLocal<WeakReference>();
+        private ConditionalWeakTable<Thread, BufferPool> _strongReferences;
+
+        public BufferPoolThreadLocal(ISerializationService serializationService)
+        {
+            _serializationService = serializationService;
+            _strongReferences = new ConditionalWeakTable<Thread, BufferPool>();
+        }
+
+        public BufferPool Get()
+        {
```

Comment:
since dispose sets the _strongReferences to null, should we do a null check here and create it if null.

Reply:
Dispose is not recoverable operation, once it is disposed, client will always shutdown and no need to recreate it.

Reply:
Java client implementation looks recoverable.

Reply:
This class itself is recoverable but it's not used as a recoverable class.  Its clear method only called from a shutdown.
Java client uses a method named `clear` to clear it, where I named it as `Dispose` and implement interface `IDisposable` to imply that this class  is not recoverable.